src/main/java/bio/terra/common/kubernetes/KubeService.java

```diff
@@ -0,0 +1,195 @@
+package bio.terra.common.kubernetes;
+
+import bio.terra.common.kubernetes.exception.KubeApiException;
+import bio.terra.stairway.Stairway;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.kubernetes.client.openapi.ApiClient;
+import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.openapi.Configuration;
+import io.kubernetes.client.openapi.apis.CoreV1Api;
+import io.kubernetes.client.openapi.models.V1Pod;
+import io.kubernetes.client.openapi.models.V1PodList;
+import io.kubernetes.client.util.ClientBuilder;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * KubeService provides access to a given Kubernetes environment.
+ *
+ * <p>The most common use in McTerra is for services that use Stairway to be able to recover flights
+ * that fail when a pod goes away. The current convention is to use the pod name as the stairway
+ * name. Then the service can compare the list of stairways it thinks should be running, compare
+ * with list of active pods, and recover anything that is no longer active.
+ */
+@SuppressFBWarnings(
+    value = "DMI_HARDCODED_ABSOLUTE_FILENAME",
+    justification = "The K8s namespace file is a valid absolute filename")
+public class KubeService {
+  private static final Logger logger = LoggerFactory.getLogger(KubeService.class);
+
+  // Location in the container where Kubernetes stores service account and
+  // namespace information. Kubernetes mints a service account for the container (pod?)
+  // that can be used to make requests of Kubernetes.
+  static final String KUBE_DIR = "/var/run/secrets/kubernetes.io/serviceaccount";
+  static final String KUBE_NAMESPACE_FILE = KUBE_DIR + "/namespace";
+
+  private final String podName;
+  private final boolean inKubernetes;
+  private final String podNameFilter;
+  private final String namespace;
+  private final AtomicBoolean isShutdown;
+
+  private KubePodListener podListener;
+  private Thread podListenerThread;
+
+  /**
+   * @param podName Name of the Kubernetes pod we are running in. If we are not in a pod, this
+   *     defaults to a constant string in application properties. We cannot easily find our own pod
+   *     name, so components must plumb this in using settings in the helm charts.
+   * @param inKubernetes Used to denote that we are running in the Kubernetes environment. The
+   *     service can be called and provides reasonable answers when inKubernetes is false.
+   * @param podNameFilter Filter to apply to pod names returned by Kubernetes API to know which pods
+   *     we care about.
+   */
+  public KubeService(String podName, boolean inKubernetes, String podNameFilter) {
```

Comment:
Your setup of `podName` and `inKubernetes` look right to me. That is what TDR does [here](https://github.com/broadinstitute/datarepo-helm/blob/master/charts/datarepo-api/templates/api-deployment.yaml)

In TDR, the podNameFilter is a constant `datarepo-api` because that substring is used  for the main DRmanager code in all deployments. So long as all Buffer Service pods in a deployment have `{{ .Values.name }}-service` as part of their name (and no other pod has it as part of its name), then it should work.